### DISCUSSION TITLE
What algorithms are used in ripgrep  ?

### DISCUSSION THREAD

**Discussion by sofeepokhrel** [BurntSushi/ripgrep#2225] [Q&A] [answered]
Hello i want to ask a different question.
Ripgrep is so damn fast , it is unbelievable that such dynamic language can beat C  .
It is impossible. 

So There must be something , architecture, algorithms used in ripgrep , that is contributing to speed factor.

Would you please give me knowledge to it?
Thank you

**Comment by BurntSushi** [comment, marked as answer]
“When you have eliminated all which is impossible, then whatever remains, however improbable, must be the truth.”

The problem is, you haven't scrutinized the assumptions you're making. For example, ripgrep is written in Rust and Rust is _not_ a dynamic language. In fact, any cursory research on Rust should lead you to the quick conclusion that one of its design goals is to be as fast as C. Rust borrows the "zero cost abstractions" design philosophy that C++ has.

> Would you please give me knowledge to it?

When I first released ripgrep, I wrote a blog post introducing it with benchmarks and a fairly detailed analysis. The [blog post](https://blog.burntsushi.net/ripgrep/) is linked in the README. Please read that first. If you still have specific questions after reading that, feel free to ask them here.

**Comment by sofeepokhrel** [comment]
Sir, Thank you for your awesome work for community.
It would be nice if ripgrep would be licensed GPL-v3 or GPL based licenses for such game changer project. :)

**Reply by BurntSushi** [reply]
Unsolicited requests to relicense a project are rude. Doubly so when it is addressed in the FAQ: https://github.com/BurntSushi/ripgrep/blob/master/FAQ.md#license
